### PR TITLE
BP should match both subdomain and path to find Service

### LINKED ISSUE #111: Border Patrol service matcher should look at both subdomain and path
Currently, BP filters by subdomain and then by path. Thus, may choose an incorrect ServiceIdentifier.

Instead, it should look at combination of subdomain & path to determine the serviceIdentifier, so that following config can work:

subdomain: foo, path: /fun
subdomain: foo, path: /sun

### PR DESCRIPTION
**Pull request by maheshkelkar**
- Fix support for multiple services with same subdomain
- Added CustomerIdentifier to represent customer specific config

Fixes #111

### CODE REVIEW

File: bpConfig.json [hunk 1]
```diff
@@ -54,18 +54,26 @@
...
       "hosts": ["http://localhost:8081"],
-      "loginManager": "dilli",
-      "subdomain": "dilli",
       "path": "/top",
       "rewritePath": "/internal/top",
-      "name": "dilli"
+      "name": "two"
+    }
+  ],
+  "customerIdentifiers": [
```
Comment (trane) [inline bpConfig.json]: curious, would it be helpful to have a `name`?

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala [hunk 2]
```diff
@@ -50,28 +49,41 @@ case class ServiceMatcher(services: Set[ServiceIdentifier]) {
...
 
   /**
-   * Derive a ServiceIdentifier from an `httpx.Request`
+   * Derive a CustomerIdentifier and ServiceIdentifier from an `httpx.Request`
    */
-  def get(req: Request): Option[ServiceIdentifier] =
-    req.host.flatMap(subdomain).filter(sid => sid.isMatchingPath(Path(req.path)))
+  def get(req: Request): Option[Tuple2[CustomerIdentifier, ServiceIdentifier]] =
+    (req.host.flatMap(subdomain), path(Path(req.path))) match {
+      case (Some(cid), Some(sid)) => Some(Tuple2(cid, sid))
```
Comment (trane) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: you don't need the `Tuple2` you can just do `Some((cid, sid))`

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala [hunk 3]
```diff
@@ -50,28 +49,41 @@ case class ServiceMatcher(services: Set[ServiceIdentifier]) {
...
+        if (sid1.path.toString.size > sid2.path.toString.size) sid1 else sid2
     )
 
   /**
-   * Derive a ServiceIdentifier from an `httpx.Request`
+   * Derive a CustomerIdentifier and ServiceIdentifier from an `httpx.Request`
    */
-  def get(req: Request): Option[ServiceIdentifier] =
-    req.host.flatMap(subdomain).filter(sid => sid.isMatchingPath(Path(req.path)))
+  def get(req: Request): Option[Tuple2[CustomerIdentifier, ServiceIdentifier]] =
```
Comment (trane) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: `def get(req: Request): Option[(CustomerIdentifier, ServiceIdentifier)]`

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala [hunk 4]
```diff
@@ -50,28 +49,41 @@ case class ServiceMatcher(services: Set[ServiceIdentifier]) {
...
+  def path(path: Path): Option[ServiceIdentifier] =
+    foldWith(
+      serviceIds.filter(sid => path.startsWith(sid.path)),
+      (sid1: ServiceIdentifier, sid2: ServiceIdentifier) =>
+        if (sid1.path.toString.size > sid2.path.toString.size) sid1 else sid2
     )
 
   /**
-   * Derive a ServiceIdentifier from an `httpx.Request`
+   * Derive a CustomerIdentifier and ServiceIdentifier from an `httpx.Request`
```
Comment (trane) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: describe the 3 cases here

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala [hunk 5]
```diff
@@ -17,10 +17,10 @@ import scala.util.{Failure, Success}
 /**
  * PODs
  */
-case class ServiceRequest(req: Request, serviceId: ServiceIdentifier)
+case class ServiceRequest(req: Request, customerId: CustomerIdentifier, serviceId: ServiceIdentifier)
```
Comment (trane) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: does it even make sense to have this distinction between `ServiceRequest` and `SessionIdRequest`? especially with that overloaded constructor on line 22.

we can remove some of those nested `req.req.req` calls if we re-signature these:

```
case class ServiceRequest(req: Request, customerId: CustomerIdentifier, serviceId: ServiceIdentifier)

case class SessionIdRequest(req: Request, customerId: CustomerIdentifier, serviceId: ServiceIdentifier, sessionId: SessionId) {
  def apply(sr: ServiceRequest, sid: SessionId): SessionIdRequest =
    this(sr.req, sr.customerId, sr.serviceId, sessionId)
}

case class AccessIdRequest[A](req: Request, customerId: CustomerIdentifier, serviceId: ServiceIdentifier, sessionId: SessionId, id: Id[A]) {
  def apply(sr: SessionIdRequest, id: Id[A]): AccessIdRequest[A] =
    this(sr.req, sr.customerId, sr.serviceId, sr.sessionId, id)
}
```

File: bpConfig.json (hunk 1 above)
Comment (maheshkelkar) [inline bpConfig.json]: no. We don't need `name`, unless of course if customer wants name to be different than subdomain.

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala (hunk 2 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: Compile complained when I did that.

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala (hunk 3 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: compiler complained. But, will try again

File: core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala (hunk 5 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/auth/BorderAuth.scala]: ok. done.

File: core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala (hunk 4 above)
Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/ServiceMatcher.scala]: done

**Comment by maheshkelkar** [comment]
@trane Let me know if you are all set with this and I will merge it. Thanks

**Pull request merged**